content/400-reference/200-api-reference/050-prisma-client-reference.mdx

```diff
@@ -5579,3 +5581,197 @@ await prisma.user.findMany({
+
+From version 4.5.0, the generated type `UserWhereUniqueInput` on [`where`](#where) exposes all fields on the model, not just unique fields. To use this feature, [enable the `extendedWhereUnique` preview flag](#enable-the-ability-to-filter-on-non-unique-fields-with-userwhereuniqueinput).
+
+You must specify at least one unique field in your `where` statement [outside of boolean operators](#boolean-operators-with-userwhereuniqueinput), and you can specify any number of additional unique and non-unique fields. You can use this to add filters to any operation that returns a single record. For example, you can use this feature for the following:
+
+- [Optimistic concurrency control on updates](#optimistic-concurrency-control-on-updates)
+- [Permission checks](#permission-checks)
+- [Soft deletes](#soft-deletes)
+
+### Enable the ability to filter on non-unique fields with <inlinecode>UserWhereUniqueInput</inlinecode>
+
+Enable the `extendedWhereUnique` preview flag in your `schema.prisma` file's `generator` block, as follows:
+
+```prisma
+generator client {
+  provider        = "prisma-client-js"
+  previewFeatures = ["extendedWhereUnique"]
+}
+```
+
+### Optimistic concurrency control on updates
+
+You can filter on non-unique fields to perform [optimistic concurrency control](/guides/performance-and-optimization/prisma-client-transactions-guide#optimistic-concurrency-control) on `update` operations.
+
+To perform optimistic concurrency control, we recommend that you use a `version` field to check whether the data in the record has changed while your code executes. Before version 4.5.0, you could not evaluate the `version` field in an `update` operation, because this field is non-unique. This is possible from version 4.5.0.
+
+In the following example, `updateOne` and `updateTwo` first read the same record and then attempt to update it. The database only executes these updates if the value in `version` is the same as the value when it did the initial read. When the database executes the first of these updates (which might be `updateOne` or `updateTwo`, depending on timing), it increments the value in `version`. This means that the database does not execute the second update because the value in `version` has changed.
+
+```prisma
+model User {
+  id      Int    @id @default(autoincrement())
+  email   String @unique
+  city    String
+  version Int
+}
+```
+
+```ts
+function updateOne() {
+   await prisma.user.update({
+     where: { id: 1, version: 0 },
+     data: { city: "Berlin", version: { increment: 1 }
+  });
+}
+
+function updateTwo() {
+   await prisma.user.update({
+     where: { id: 1, version: 0 },
+     data: { city: "New York", version: { increment: 1 }
+  });
+}
+
+function main () {
+  await Promise.allSettled([updateOne(), updateTwo()]);
+}
+```
+
+<Admonition type="info">
+
+In optimistic concurrency control, you must fetch the record, relation, or nested record before you update it.
+
+**Have I mangled this with my rewrite?**
```

Comment:
No you haven't but you might want to add this in your example above.

```
function updateOne() {
  const user = await prisma.user.findUnique({ id: 1 });

   await prisma.user.update({
     where: { id: user.id, version: user.version },
     data: { city: "Berlin", version: { increment: 1 }
  });
}
```

Reply:
Just to make sure you understand why you need to fetch the record before:

In your example, you're querying for `version: 0`. That's ok as an example but in a real world use-case, you wouldn't know what the version is. So, to know that version, you have to query the user first and use whatever version it's at, and then increment it. That's why you need to fetch before. It's not so much that OCC needs it, but rather that you need that info and the only way to get it is to fetch it first.

The more I read this note the more I'm starting to think we should just remove it and just have the example fetch the record before doing the update just like on my example above.

Reply:
Thanks - that makes the example code much better.